File: agent/src/hidpp/transport.cpp
#include "transport.h"

#include <fcntl.h>
#include <linux/hidraw.h>
#include <poll.h>
#include <sys/ioctl.h>
#include <unistd.h>

#include <chrono>
#include <cstring>

namespace hidpp {

HidppError::HidppError(uint8_t c, uint8_t f, uint8_t fn)
    : std::runtime_error("HID++ error 0x" + std::to_string(c) + " (feature idx " + std::to_string(f) + ")"),
      code(c), featureIndex(f), function(fn) {}

RawInfo rawInfo(const std::string& path) {
    RawInfo r;
    int fd = ::open(path.c_str(), O_RDONLY | O_NONBLOCK);
    if (fd < 0) throw std::runtime_error("open " + path + ": " + strerror(errno));
    hidraw_devinfo di{};
    if (ioctl(fd, HIDIOCGRAWINFO, &di) == 0) {
        r.bustype = di.bustype;
        r.vendor = static_cast<uint16_t>(di.vendor);
        r.product = static_cast<uint16_t>(di.product);
    }
    char name[256] = {0};
    if (ioctl(fd, HIDIOCGRAWNAME(sizeof(name)), name) >= 0) r.name = name;
    ::close(fd);
    return r;
}

bool supportsHidpp(const std::string& path) {
    int fd = ::open(path.c_str(), O_RDONLY | O_NONBLOCK);
    if (fd < 0) return false;
    int size = 0;
    bool ok = false;
    if (ioctl(fd, HIDIOCGRDESCSIZE, &size) == 0 && size > 0) {
        hidraw_report_descriptor rd{};
        rd.size = size;
        if (ioctl(fd, HIDIOCGRDESC, &rd) == 0) {
            const uint8_t a[] = {0x06, 0x00, 0xff}, b[] = {0x06, 0x43, 0xff};
            for (int i = 0; i + 3 <= size && !ok; ++i)
                ok = !memcmp(rd.value + i, a, 3) || !memcmp(rd.value + i, b, 3);
        }
    }
    ::close(fd);
    return ok;
}

Transport::Transport(std::string path, Callback cb, double timeoutSec)
    : path_(std::move(path)), timeout_(timeoutSec), cb_(std::move(cb)) {
    info_ = rawInfo(path_);
    fd_ = ::open(path_.c_str(), O_RDWR | O_NONBLOCK);
    if (fd_ < 0) throw std::runtime_error("open " + path_ + ": " + strerror(errno));
    thread_ = std::thread([this] { reader(); });
}

void Transport::setCallback(Callback cb) {
    std::lock_guard<std::mutex> lk(cbMutex_);
    cb_ = std::move(cb);
}

Transport::~Transport() {
    stop_ = true;
    if (thread_.joinable()) thread_.join();
    if (fd_ >= 0) ::close(fd_);
}

void Transport::reader() {
    pollfd p{fd_, POLLIN, 0};
    uint8_t buf[64];
    while (!stop_) {
        int r = ::poll(&p, 1, 200);
        if (r <= 0) continue;
        if (p.revents & (POLLERR | POLLHUP | POLLNVAL)) return;
        ssize_t n = ::read(fd_, buf, sizeof(buf));
        if (n <= 0) continue;
        dispatch(Bytes(buf, buf + n));
    }
}

void Transport::dispatch(const Bytes& d) {
    if (d.size() < 4) return;
    uint8_t rid = d[0];
    if (rid != kShort && rid != kLong && rid != kVeryLong) return;
    uint8_t dev = d[1], b2 = d[2], b3 = d[3];
    {
        std::lock_guard<std::mutex> lk(slotMutex_);
        if (pending_ && dev == pDev_) {
            if ((b2 == 0x8F || b2 == 0xFF) && d.size() >= 6) {
                // error frame: [rid dev 0x8F/0xFF featIdx fn code]
                uint8_t ef = d[3], efn = d[4], code = d[5];
                bool match = (ef == pB2_ && (pMatchSw_ ? (efn & 0xF0) == (pB3_ & 0xF0) : efn == pB3_)) ||
                             (b2 == 0x8F && (efn & 0xF0) == (pB3_ & 0xF0));
                if (match) {
                    error_ = HidppError(code, ef, efn);
                    haveResult_ = true;
                    pending_ = false;
                    cv_.notify_all();
                    return;
                }
            } else if (b2 == pB2_) {
                bool match = pMatchSw_ ? (b3 == pB3_) : (b3 == pB3_);
                if (match) {
                    result_.assign(d.begin() + 4, d.end());
                    haveResult_ = true;
                    pending_ = false;
                    cv_.notify_all();
                    return;
                }
            }
        }
    }
    if (b2 == 0x8F || b2 == 0xFF) return;
    Callback cb;
    {
        std::lock_guard<std::mutex> lk(cbMutex_);
        cb = cb_;
    }
    if (cb) {
        Notification n;
        n.deviceIndex = dev;
        n.featureIndex = b2;
        n.event = b3 >> 4;
        n.swId = b3 & 0x0F;
        n.address = b3;
        n.reportId = rid;
        n.data.assign(d.begin() + 4, d.end());
        try {
            cb(n);
        } catch (...) {
        }
    }
}

Bytes Transport::exchange(const Bytes& frame, uint8_t devIdx, uint8_t b2, uint8_t b3, bool matchSw, bool noReply) {
    std::lock_guard<std::mutex> req(reqMutex_);
    {
        std::lock_guard<std::mutex> lk(slotMutex_);
        pending_ = !noReply;
        pDev_ = devIdx;
        pB2_ = b2;
        pB3_ = b3;
        pMatchSw_ = matchSw;
        haveResult_ = false;
        error_.reset();
        result_.clear();
    }
    ssize_t w = ::write(fd_, frame.data(), frame.size());
    if (w < 0) {
        std::lock_guard<std::mutex> lk(slotMutex_);
        pending_ = false;
        throw std::runtime_error(std::string("write: ") + strerror(errno));
    }
    if (noReply) return {};
    std::unique_lock<std::mutex> lk(slotMutex_);
    if (!cv_.wait_for(lk, std::chrono::duration<double>(timeout_), [this] { return haveResult_; })) {
        pending_ = false;
        throw Timeout("no reply");
    }
    if (error_) throw *error_;
    return result_;
}

Bytes Transport::request(uint8_t devIdx, uint8_t featIdx, uint8_t fn, const Bytes& params,
                         std::optional<bool> longReport, bool noReply) {
    bool lng = longReport.value_or(params.size() > 3 || info_.bustype == 0x05);
    uint8_t fnsw = static_cast<uint8_t>(((fn & 0x0F) << 4) | kSwId);
    Bytes f = {lng ? kLong : kShort, devIdx, featIdx, fnsw};
    f.insert(f.end(), params.begin(), params.end());
    f.resize(lng ? kLongLen : kShortLen, 0);
    return exchange(f, devIdx, featIdx, fnsw, true, noReply);
}

Bytes Transport::request10(uint8_t devIdx, uint8_t subId, uint8_t reg, const Bytes& params, bool longReport) {
    Bytes f = {longReport ? kLong : kShort, devIdx, subId, reg};
    f.insert(f.end(), params.begin(), params.end());
    f.resize(longReport ? kLongLen : kShortLen, 0);
    return exchange(f, devIdx, subId, reg, false, false);
}

std::optional<std::pair<int, int>> Transport::ping(uint8_t devIdx) {
    try {
        Bytes r = request(devIdx, 0x00, 0x1, {0, 0, 0x5a}, false);
        if (r.size() < 2) return std::nullopt;
        return std::make_pair(r[0], r[1]);
    } catch (const HidppError&) {
        return std::make_pair(1, 0);
    } catch (const std::exception&) {
        return std::nullopt;
    }
}

std::optional<PairingInfo> Transport::pairingInfo(uint8_t idx) {
    try {  // Bolt: 0xB5 / 0x50 + idx
        Bytes r = request10(0xFF, 0x83, 0xB5, {static_cast<uint8_t>(0x50 + idx)});
        if (r.size() >= 8 && r[0] == 0x50 + idx) {
            PairingInfo p;
            p.wpid = static_cast<uint16_t>((r[3] << 8) | r[2]);
            p.kind = r[1] & 0x0F;
            char s[9];
            snprintf(s, sizeof(s), "%02X%02X%02X%02X", r[4], r[5], r[6], r[7]);
            p.serial = s;
            return p;
        }
    } catch (const std::exception&) {
    }
    try {  // Unifying / Nano: 0xB5 / 0x20 | (idx-1)
        Bytes r = request10(0xFF, 0x83, 0xB5, {static_cast<uint8_t>(0x20 | (idx - 1))});
        if (r.size() >= 8 && (r[0] & 0x0F) == idx - 1) {
            PairingInfo p;
            p.wpid = static_cast<uint16_t>((r[3] << 8) | r[4]);
            p.kind = r[7] & 0x0F;
            return p;
        }
    } catch (const std::exception&) {
    }
    return std::nullopt;
}

}  // namespace hidpp

File: logimx/hidpp/transport.py
"""hidraw transport for HID++ 1.0/2.0 receivers and devices.

One Transport owns one /dev/hidrawN node (a receiver, or a directly connected
Bluetooth / USB device). Requests are serialised; notifications are delivered
to a callback from the reader thread.
"""
from __future__ import annotations

import fcntl
import os
import queue
import select
import struct
import threading
import time
from dataclasses import dataclass
from typing import Callable, Optional

SHORT = 0x10   # 7 bytes
LONG = 0x11    # 20 bytes
VERY_LONG = 0x12  # 64 bytes
DJ_SHORT = 0x20
DJ_LONG = 0x21

SHORT_LEN = 7
LONG_LEN = 20

SW_ID = 0x0D  # our software id (low nibble of the function byte); Solaar uses another

# HIDIOCGRAWINFO / HIDIOCGRAWNAME(len) from linux/hidraw.h
_HIDIOCGRAWINFO = 0x80084803
_HIDIOCGRAWNAME = lambda n: 0x80004804 | (n << 16)  # noqa: E731
_HIDIOCGRDESCSIZE = 0x80044801
_HIDIOCGRDESC = 0x90044802


@dataclass
class RawInfo:
    bustype: int
    vendor: int
    product: int
    name: str


def raw_info(path: str) -> RawInfo:
    fd = os.open(path, os.O_RDONLY | os.O_NONBLOCK)
    try:
        buf = fcntl.ioctl(fd, _HIDIOCGRAWINFO, bytes(8))
        bustype, vendor, product = struct.unpack("<IHH", buf)
        name = fcntl.ioctl(fd, _HIDIOCGRAWNAME(256), bytes(256)).split(b"\0", 1)[0].decode(errors="replace")
        return RawInfo(bustype, vendor & 0xFFFF, product & 0xFFFF, name)
    finally:
        os.close(fd)


def report_descriptor(path: str) -> bytes:
    fd = os.open(path, os.O_RDONLY | os.O_NONBLOCK)
    try:
        size = struct.unpack("<I", fcntl.ioctl(fd, _HIDIOCGRDESCSIZE, bytes(4)))[0]
        buf = bytearray(4 + 4096)
        struct.pack_into("<I", buf, 0, size)
        fcntl.ioctl(fd, _HIDIOCGRDESC, buf, True)
        return bytes(buf[4:4 + size])
    finally:
        os.close(fd)


def supports_hidpp(path: str) -> bool:
    """True if the report descriptor exposes the vendor usage pages
    used by HID++ (0xFF00 / 0xFF43 with report ids 0x10 / 0x11)."""
    try:
        desc = report_descriptor(path)
    except OSError:
        return False
    return b"\x06\x00\xff" in desc or b"\x06\x43\xff" in desc


class HidppError(Exception):
    def __init__(self, code: int, feature_index: int = 0, function: int = 0):
        self.code = code
        self.feature_index = feature_index
        self.function = function
        super().__init__(f"HID++ error 0x{code:02x} (feature idx 0x{feature_index:02x}, fn 0x{function >> 4:x})")


class Timeout(Exception):
    pass


@dataclass
class Notification:
    device_index: int
    feature_index: int  # for HID++ 2.0: feature index; for 1.0: sub id
    event: int          # high nibble of byte 3 (function/event id)
    sw_id: int
    data: bytes
    report_id: int


class Transport:
    def __init__(self, path: str, on_notification: Optional[Callable[[Notification], None]] = None,
                 timeout: float = 1.5):
        self.path = path
        self.info = raw_info(path)
        self.fd = os.open(path, os.O_RDWR | os.O_NONBLOCK)
        self.timeout = timeout
        self._lock = threading.Lock()
        self._pending: dict[tuple, queue.Queue] = {}
        self._pending_lock = threading.Lock()
        self.on_notification = on_notification
        self._stop = threading.Event()
        self._thread = threading.Thread(target=self._reader, name=f"hidpp-reader-{os.path.basename(path)}", daemon=True)
        self._thread.start()

    # ------------------------------------------------------------------ io
    def close(self):
        self._stop.set()
        try:
            os.close(self.fd)
        except OSError:
            pass

    def _reader(self):
        while not self._stop.is_set():
            try:
                r, _, _ = select.select([self.fd], [], [], 0.5)
            except (OSError, ValueError):
                return
            if not r:
                continue
            try:
                data = os.read(self.fd, 64)
            except BlockingIOError:
                continue
            except OSError:
                return
            if not data:
                continue
            self._dispatch(data)

    def _dispatch(self, data: bytes):
        rid = data[0]
        if rid not in (SHORT, LONG, VERY_LONG):
            return  # DJ reports and plain HID reports are not ours
        if len(data) < 4:
            return
        devidx, feat, fnsw = data[1], data[2], data[3]
        payload = data[4:]
        # HID++ 1.0 error: sub id 0x8F ; HID++ 2.0 error: feature index 0xFF
        if feat == 0x8F or feat == 0xFF:
            err_feat, err_fn, err_code = data[3], data[4], data[5]
            if not self._resolve((devidx, err_feat, err_fn & 0xF0), HidppError(err_code, err_feat, err_fn)):
                self._resolve((devidx, 0x8F, err_fn & 0xF0), HidppError(err_code, err_feat, err_fn))
            return
        key = (devidx, feat, fnsw & 0xF0)
        if feat in (0x80, 0x81, 0x82, 0x83):  # HID++ 1.0 register reply: byte3 is the register
            if self._resolve((devidx, feat, fnsw), data[4:]):
                return
        elif (fnsw & 0x0F) == SW_ID and self._resolve(key, payload):
            return
        # everything else (sw id 0, or unmatched) is a notification/event
        if self.on_notification:
            try:
                self.on_notification(Notification(devidx, feat, fnsw >> 4, fnsw & 0x0F, payload, rid))
            except Exception:  # never kill the reader
                pass

    def _resolve(self, key, value) -> bool:
        with self._pending_lock:
            q = self._pending.get(key)
        if q is None:
            return False
        q.put(value)
        return True

    # ------------------------------------------------------------- request
    def request(self, device_index: int, feature_index: int, function: int, params: bytes = b"",
                long: Optional[bool] = None, no_reply: bool = False, timeout: Optional[float] = None) -> bytes:
        """Send a HID++ 2.0 request. `function` is the function number (0..15).
        Returns the payload (16 bytes for long replies)."""
        fnsw = ((function & 0x0F) << 4) | SW_ID
        if long is None:
            long = len(params) > 3 or self.info.bustype == 0x05  # BT prefers long reports
        if long:
            frame = struct.pack("BBBB", LONG, device_index, feature_index, fnsw) + params.ljust(LONG_LEN - 4, b"\0")
        else:
            frame = struct.pack("BBBB", SHORT, device_index, feature_index, fnsw) + params.ljust(SHORT_LEN - 4, b"\0")
        key = (device_index, feature_index, fnsw & 0xF0)
        q: queue.Queue = queue.Queue()
        with self._lock:
            with self._pending_lock:
                self._pending[key] = q
            try:
                os.write(self.fd, frame)
                if no_reply:
                    return b""
                try:
                    res = q.get(timeout=timeout or self.timeout)
                except queue.Empty:
                    raise Timeout(f"no reply from device {device_index} feature 0x{feature_index:02x} fn {function}")
            finally:
                with self._pending_lock:
                    self._pending.pop(key, None)
        if isinstance(res, Exception):
            raise res
        return res

    def request10(self, device_index: int, sub_id: int, register: int, params: bytes = b"",
                  long: bool = False, timeout: Optional[float] = None) -> bytes:
        """HID++ 1.0 register access (sub_id 0x80 set / 0x81 get / 0x82 set long / 0x83 get long)."""
        rid, size = (LONG, LONG_LEN) if long else (SHORT, SHORT_LEN)
        frame = struct.pack("BBBB", rid, device_index, sub_id, register) + params.ljust(size - 4, b"\0")
        reply_sub = sub_id  # replies echo the sub id
        key = (device_index, reply_sub, register)
        q: queue.Queue = queue.Queue()
        with self._lock:
            with self._pending_lock:
                self._pending[key] = q
                self._pending[(device_index, 0x8F, register & 0xF0)] = q
            try:
                os.write(self.fd, frame)
                try:
                    res = q.get(timeout=timeout or self.timeout)
                except queue.Empty:
                    raise Timeout(f"no reply to register 0x{register:02x}")
            finally:
                with self._pending_lock:
                    self._pending.pop(key, None)
                    self._pending.pop((device_index, 0x8F, register & 0xF0), None)
        if isinstance(res, Exception):
            raise res
        return res

    def pairing_info(self, device_index: int) -> Optional[tuple[int, int, str]]:
        """(wpid, device_kind, serial) for a receiver slot. Bolt: register 0xB5 sub 0x50+idx;
        Unifying/Nano: 0xB5 sub 0x20+idx-1."""
        try:
            r = self.request10(0xFF, 0x83, 0xB5, bytes([0x50 + device_index]))
            if len(r) >= 8 and r[0] == 0x50 + device_index:
                return (r[3] << 8) | r[2], r[1] & 0x0F, r[4:8].hex().upper()
        except (HidppError, Timeout, OSError):
            pass
        try:
            r = self.request10(0xFF, 0x83, 0xB5, bytes([0x20 | (device_index - 1)]))
            if len(r) >= 8 and (r[0] & 0x0F) == device_index - 1:
                return struct.unpack(">H", r[3:5])[0], r[7] & 0x0F, ""
        except (HidppError, Timeout, OSError):
            pass
        return None

    def ping(self, device_index: int) -> Optional[tuple[int, int]]:
        """ROOT.getProtocolVersion. Returns (major, minor) or None if the device is
        not reachable (HID++ 1.0 devices answer with an error 0x8F)."""
        try:
            r = self.request(device_index, 0x00, 0x1, b"\0\0\x5a", long=False)
        except HidppError:
            return (1, 0)
        except (Timeout, OSError):
            return None
        return (r[0], r[1])

    def __repr__(self):
        return f"Transport({self.path}, {self.info.vendor:04x}:{self.info.product:04x} '{self.info.name}')"


def _sleep(s: float):
    time.sleep(s)
